src/internal/puck/components/Header.tsx

```diff
@@ -97,6 +97,7 @@ export const customHeader = (
             onClick={async () => {
               await handleSaveData(currentPuckData);
               handleClearLocalChanges();
+              setHistories([{ id: "root", state: { data: currentPuckData } }]);
```

Comment:
I think what you really want to do is re-setPuckInitialHistory in clearHistory func, or maybe setPuckInitialHistoryFetched(false) so that loadPuckInitialHistory gets rerun

Reply:
I was thinking about that, but I feel like re-setting puckInitialHistory goes against what the initial history is meant to be. I don't think we should have two history objects that we update separately as it was already pretty confusing while trying to debug this.

When we clear local changes we update the histories to match the first entry. Before we were using puckInitialHistory, but I think that should just be using puck history directly. I've tested and this setup works consistently and stays synchronized.

We also didn't add updating the histories for saving, which I've added here.

Reply:
puckInitialHistory is the one history. What do you mean in your first comment?
> re-setting puckInitialHistory goes against what the initial history is meant to be
> I don't think we should have two history objects that we update separately

puckInitialHistory is used to tell Puck what history state (and histories) to render. What are we going against? And what is the second history object you're referring to?

If we do the work in clearHistory it handles both clearLocalChanges and publish since both of those call clearHistory.

Reply:
In most other places outside of initialization, we are using the actual history/setHistories from usePuck. I think we should use that instead to keep things consistent, because this problem was caused by a desync between the two.

I'll move the change over to handleClearLocalChanges. We also have to pass in the data to set the history to.